src/AggregateFunctions/AggregateFunctionMinMaxAny.h

```diff
@@ -506,34 +513,40 @@ struct SingleValueDataString //-V730
         Int32 rhs_size;
         readBinary(rhs_size, buf);
 
-        if (rhs_size >= 0)
+        if (rhs_size <= 0)
         {
-            if (rhs_size <= MAX_SMALL_STRING_SIZE)
-            {
-                /// Don't free large_data here.
-
-                size = rhs_size;
-
-                if (size > 0)
-                    buf.read(small_data, size);
-            }
-            else
-            {
-                if (capacity < rhs_size)
-                {
-                    capacity = static_cast<UInt32>(roundUpToPowerOfTwoOrZero(rhs_size));
-                    /// Don't free large_data here.
-                    large_data = arena->alloc(capacity);
-                }
+            size = 1;
+            small_data[0] = '\0';
+            using_internal = true;
+            return;
+        }
 
-                size = rhs_size;
-                buf.read(large_data, size);
-            }
+        if (rhs_size <= MAX_SMALL_STRING_SIZE)
+        {
+            size = rhs_size;
+            buf.read(small_data, size);
+            using_internal = true;
         }
         else
         {
-            /// Don't free large_data here.
+            Int32 size_with_possible_null = rhs_size + 1;
+            if (capacity < size_with_possible_null)
+            {
+                capacity = static_cast<UInt32>(roundUpToPowerOfTwoOrZero(size_with_possible_null));
+                large_data = arena->alloc(capacity);
+            }
+
             size = rhs_size;
+            buf.read(large_data, size);
+            using_internal = false;
+        }
+
+        char * data = getData();
+        if (data[size - 1] != '\0')
```

Comment:
Nice catch. I'm not sure if there is a way to workaround this. I'll need to think on it.

> How does that sounds to you?

I think we need something that can be backported, and backporting a new serialization version doesn't seem like a good idea to me. I also have my reservations about how the serialization version really works on cluster when using different CH releases (i.e. during cluster upgrades), but it's mostly because I haven't really looked deep into it.


Having a warning makes sense, but I'm unsure how to do it. Should we warn once per string with issues(too many), once per class (too many too), once per query (requires synchronization and access to the context)? Maybe it'd be possible to do it during table load, but then it requires checking all the data during table load and that seems it'd take quite a bit of work and might be "very bad for performance". Maybe there is a better way to warn about this directly in the process without being too noisy or too complex.

Reply:
>I think we need something that can be backported, and backporting a new serialization version doesn't seem like a good idea to me

The whole situation is not good, so I think it is fine to use different aggregation states version, but maybe people from ClickHouse inc has other opinion... @alexey-milovidov any thoughts on this one?

>I also have my reservations about how the serialization version really works on cluster when using different CH releases

Even if there are some issues it should be better then ignoring NULL at the end of strings.

>Maybe it'd be possible to do it during table load, but then it requires checking all the data during table load and that seems it'd take quite a bit of work

I thought about adding a warning for the table, version for the aggregation states works based on revision, so this patch should increment the revision (it is also not great to backport this patch, but maybe it is OK for this situation) and if there is `AggregateFunction(argMax, String, ...)` (and all others that uses this code) and version is `0` then log a warning.

Yes, this looks like a hack, but I can't came up with better idea.

Reply:
My concern with switching to a different aggregate version is that it's a mayor breaking change that needs a lot of supervision.

Let's say I'm running 4 servers of 22.8.5 (without the bug) and I want to upgrade to 22.8.10 (with the new version). What happens the moment I upgrade a replica and that replica writes the new states? The other 3 replicas won't know how to read them and, best case scenario they fail, the second best case is that they output bad data, and the worse case is that they crash. I don't see hard checks to control versioning (on old releases), but as I've said I'm not familiar with that code.

Reply:
>Let's say I'm running 4 servers of 22.8.5 (without the bug) and I want to upgrade to 22.8.10 (with the new version). What happens the moment I upgrade a replica and that replica writes the new states? The other 3 replicas won't know how to read them and, best case scenario they fail, the second best case is that they output bad data, and the worse case is that they crash. I don't see hard checks to control versioning (on old releases).

I see you point, indeed there are some tricky moments. Looks like to make it work correctly it is enough for new state be compatible with the old one, so new state simply should not add NULL byte at the end conditionally (and should include good comment).

Reply:
This is really tricky because yes, my main target is that the new state should work exactly the same as the old one, but still should be able to read the buggy one without major issues.


I currently don't see a good way to handle strings ending in NULL. Let's see it with an example:
* Let's say we have "0123"
  - Used to be stored as size 5. Stored as "0123\0". Read as size 5 from disk so in memory you have "0123\0".
  - It was changed to be stored as size 4. Stored in disk also as "0123\0". Read as size 4 from disk, so in memory you have "0123".

* Now we have "0123\0":
  - Used to be stored as size 6. Stored as "0123\0\0". Read a size 6 from disk so in memory you have "0123\0\0".
  - With the changes it was stored as size 5. Disk "0123\0\0". Read as size 5, so in memory you have "0123\0".

Since the current approach does not read NULLs, effectively reading an old state with "0123" and reading a buggy state with "0123\0" are the same (points 1 and 4). Both are marked as size 5 and both are represented the same in memory so I don't know if there is another (safe) way to distinguish between the 2.

Reply:
>I don't know if there is another (safe) way to distinguish between the 2.

Only if you will have some version of the format, and this is what versioned aggregation states for.

Right now there is only one version - `0`, and it is buggy, it should be interpreted with special care, like your patch adds, however it is not lossless transformation - it does not interpret strings with NULL at the end correctly.

So what I'm suggesting is, keep conditional NULL byte handling (as your patch does) for version `0` only, and introduce new version - `1`, that should not do this conditional handling, that way newly written data will be handled correctly with and without NULL at the end of string.

As for how it works with servers with different versions, there is a protocol that should serialize data for requested version (there is a even a test for this - `tests/integration/test_version_update/test.py`), so everything should be safe, but of course this should be checked.

Though maybe it should be done separately...

P.S. If you want I can submit a PR that will show the idea.

Reply:
What about data shared in replication? Does is handle different versions too?

Reply:
For the replication nothing will be changed since for existing tables nothing will be changed it will use old version.

Reply:
I mean, if I have 4 servers and update 1. What happens if this new server ingest data or merges parts (1)? Will the data created with the new version of the aggregation be replicated as-is to the others servers or is the replication system intelligent enough to rewrite the statuses in the old version for the older replicas?

* (1) Note that when merging parts, depending on the settings, either the part will be downloaded by the other replicas or other parts will do the merge themselves and the checksums won't match.

Reply:
I've been doing some tests and the situation above works because the stored type is implicitly declaring the version of the aggregation (`AggregateFunction(argMax, String, UInt64)` != `AggregateFunction(1, argMax, String, UInt64)`). So if the table has been already created then any inserts, either in the old server or the new, should use the old serialization method (version=1).

OTOH, this is broken if you create a new table on a cluster and the creation lands in the new server, as the table will be created as `AggregateFunction(1, argMax, String, UInt64)` which the old server will happily work with and might crash if the data is not compatible. So when introducing a versioned status you'd need to make sure any DDL does not land in a new server if there are any old servers still in the cluster. It's also a backwards incompatible change once applied downgrading is likely impossible.

Based on this I think it would indeed make sense to first fix the bug in a backport'able manner, and then introduce a new status version in the next stable release only (22.11 or 22.12). Since that change would affect quite a bit of aggregation functions this would require to tread lightly, lots of tests and big warnings in the announcement.

Reply:
>I've been doing some tests and the situation above works because the stored type is implicitly declaring the version of the aggregation (AggregateFunction(argMax, String, UInt64) != AggregateFunction(1, argMax, String, UInt64)). So if the table has been already created then any inserts, either in the old server or the new, should use the old serialization method (version=1).

Indeed!

>OTOH, this is broken if you create a new table on a cluster and the creation lands in the new server, as the table will be created as AggregateFunction(1, argMax, String, UInt64) which the old server will happily work with and might crash if the data is not compatible. So when introducing a versioned status you'd need to make sure any DDL does not land in a new server if there are any old servers still in the cluster.

In general - yes. But for this particular case I'm suggesting to make the underlying state the same, versions is required to avoid interpreting NULL at then end of String specially. 

>It's also a backwards incompatible change once applied downgrading is likely impossible.

I would not recommend to downgrade DBMS anyway (thought I guess it OK for minor upgrades) :)

Reply:
But I agree with you that this can be done separately.

>The definition of the column is different between the 2 servers!!!!:

Yep, and even if `argMax(String, UInt64)` is not that common, it will make the difference for all types.

Reply:
Keep in mind that it affects many other functions too.

But I've been thinking and I think the general state version change should be doable in a more upgrade friendly way by introducing a server setting with the default value. We introduce the the setting with the value set to the old(0) and a big system.warning if at the start the value is still 0. Then you can update the servers and once all are compatible change the settings. After a few stable releases we change the default to 1 announcing the incompatibility in the changelog. After N releases we can deprecate the setting and reading/writting old versions if maintenance is too much. Is not immediate but it makes updates workable

Reply:
With the idea above there would be a timeframe were you are still vulnerable to different servers creating different types while all get their settings changed. I think that could be solved by changing the definition before sending it to Keeper to contain the final version instead of the default, so all servers create the same table structure independently from their settings.

WDYT @tavplubix. Would a process like that make sense to keep compatibility and make upgrades "painless"?